### PR TITLE
member cli: add -filter expression to flags

### PR DESCRIPTION
**Pull request by huikang**
### Description

Add `-filter` to `consul members` to filter the members by bexpr. Please see https://github.com/hashicorp/consul/issues/18152 for the use case

- unit test added to api

#### How to Test

**Without filter**: all members from federated clusters will be listed:
```
consul members -wan                            
Node          Address              Status  Type    Build      Protocol  DC   Partition  Segment
server-1.dc2  192.168.1.168:8322   alive   server  1.17.0dev  2         dc2  default    <all>
server-1.dc3  192.168.1.168:28322  alive   server  1.17.0dev  2         dc3  default    <all>
```

**Filtered results**: only list the members from a specified dc

```
consul members -wan -filter 'Tags["dc"] == dc2'           
Node          Address             Status  Type    Build      Protocol  DC   Partition  Segment
server-1.dc2  192.168.1.168:8322  alive   server  1.17.0dev  2         dc2  default    <all>


consul members -wan -filter 'Tags["dc"] == dc3'
Node          Address              Status  Type    Build      Protocol  DC   Partition  Segment
server-1.dc3  192.168.1.168:28322  alive   server  1.17.0dev  2         dc3  default    <all>

```

-->

### Links

Fix https://github.com/hashicorp/consul/issues/18152

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

**Comment by huikang** [comment]
@david-yu , please let me know if the change makes sense and needs backport.

**Comment by david-yu** [comment]
Makes sense to backport but we probably should have our docs updated to reflect this new flag. Thank you.

**Comment by huikang** [comment]
> Makes sense to backport but we probably should have our docs updated to reflect this new flag. Thank you.

Sure, let me update the doc as well.

**Comment by david-yu** [comment]
Added backports to 1.16.x, 1.15.x, and 1.14.x

File: api/agent_test.go [hunk 1]
```diff
@@ -155,6 +155,16 @@ func TestAPI_AgentMembersOpts(t *testing.T) {
...
+
+	members, err = agent.MembersOpts(MembersOpts{
+		WAN:    true,
+		Filter: `Tags["dc"] == dc2`,
+	})
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	require.Equal(t, 1, len(members))
```
Comment (skpratt) [inline api/agent_test.go:167]: could we test the nil result case here too? 
i.e. Tags["dc"] == doesNotExist

Comment (huikang) [reply api/agent_test.go:167]: @skpratt , that's a good point. More test cases are added. Thanks.

**Pull request merged**